freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py

```diff
@@ -167,3 +179,152 @@ def _filter_images_with_higher_rpm_nvr(self, event, rebuild_images_list):
+        :rtype: tuple
+        """
+        renewed_odcs_compose_set = set()
+        new_odcs_compose_set = set()
+        for compose_id in original_odcs_compose_ids:
+            compose = self.odcs.get_compose(compose_id)
+            source_type = compose.get("source_type")
+            # source_type of module composes is 2
+            if source_type != 2:
+                renewed_odcs_compose_set.add(compose_id)
+                continue
+
+            name_stream_set = set()
+            source_value = compose.get("source", "")
+            for source in source_value.split():
+                group = source.strip().split(":")[:2]
+                name_stream_set.add(group[0] + ":" + group[1])
+
+            # If all the sources are not in advisory, renew them
+            if len(name_stream_set - module_name_stream_set) == len(name_stream_set):
+                renewed_odcs_compose_set.add(compose_id)
+            else:
+                # If some source in advisory, create composes for all separately
+                new_odcs_compose_set.update(name_stream_set)
+            # All modules in advisory should create compose
+            new_odcs_compose_set.update(module_name_stream_set)
+
+        return renewed_odcs_compose_set, new_odcs_compose_set
+
+    def _record_builds(self, images, event):
+        """
+        Records the images to database.
+
+        :param images list: ContainerImage instances.
+        :param event FlatpakModuleAdvisoryReadyEvent: The event this handler
+            is currently handling.
+        :return: a mapping between docker image build NVR and
+            corresponding ArtifactBuild object representing a future rebuild of
+            that docker image. It is extended by including those docker images
+            stored into database.
+        :rtype: dict
+        """
+        db_event = Event.get_or_create_from_event(db.session, event)
+
+        # Cache for ODCS module composes. Key is white-spaced, sorted, list
+        # of module's NAME:STREAM. Value is Compose database object.
+        odcs_cache = {}
+
+        # Used as tmp dict with {brew_build_nvr: ArtifactBuild, ...} mapping.
+        builds = {}
+        for image in images:
+            self.set_context(db_event)
+
+            image.resolve_commit()
+            image.resolve_flatpak_odcs_compose_ids()
+            nvr = image.nvr
+            image_name = koji.parse_NVR(nvr)["name"]
+            build = self.record_build(
+                event,
+                image_name,
+                ArtifactType.IMAGE,
+                state=ArtifactBuildState.PLANNED.value,
+                original_nvr=nvr,
+                rebuild_reason=RebuildReason.DIRECTLY_AFFECTED.value,
+            )
+            # Set context to particular build so logging shows this build
+            # in case of error.
+            self.set_context(build)
+
+            module_name_stream_set = self._get_module_name_stream_set(event)
```

Comment:
For every image, you're calling this API and the underline errata API once which is unnecessary.

Reply:
Sure. Fixing.